README.md
# spotify-preview
AWS Lambda function to get Spotify song's metadata

## First steps in development environment
1. `chmod +x *.sh` – set execute permission on bash scripts for your user
2. `source initproject.sh` – create virtualenv and install dependencies.
3. `activate` – activate virtualenv
4. `make test` – run test to check if everything works (or just `./test.sh` if you don't have `make`)

docs/CHANGELOG.md
# Changelog

_Last modification: **17.03.2022**_

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- Static analysis tools: black, flake8, isort
- Type checking with MyPy
- README
- Makefile with some helper commands

### Changed
- Reformatted files according to black, flake8 and isort rules
- Added type annotations to all functions

## [1.0.0] – 2022-04-17
### Added
* First, fully functioning version of spotify-preview. Features simple metadata retrieval from Spotify song.
* Bash scripts to build ZIPs, initialize or clear project
* Docs

test.py
import json
import unittest
from copy import deepcopy
from typing import Any, Dict
from unittest import mock
from unittest.mock import MagicMock

from requests import Response

from main import (NoDataError, ValidationError, build_track_url,
                  get_meta_props, get_response, get_song_data, get_url,
                  handle_error, lambda_handler, validate_spotify_url)
from test_data import (API_GATEWAY_EVENT, EXPECTED_META_PROPS,
                       SPOTIFY_TRACK_HTML_CONTENT,
                       SPOTIFY_TRACK_HTML_CONTENT_WITHOUT_META)


def make_api_gateway_event(url: str) -> Dict[str, Any]:
    event = deepcopy(API_GATEWAY_EVENT)
    event["queryStringParameters"]["url"] = url
    event["multiValueQueryStringParameters"]["url"] = [url]
    return event


def raise_error(error: Exception):
    raise error


class TestValidateSpotifyUrl(unittest.TestCase):
    """Test :py:func:`main.validate_spotify_url`"""

    def test_valid(self) -> None:
        track_id = "1UUBkjBgce8P53hQC3nHtU"
        valid_identifiers = [
            f"https://open.spotify.com/track/{track_id}?si=76e0e46d5d7742e0",
            f"https://open.spotify.com/track/{track_id}?si=",
            f"http://open.spotify.com/track/{track_id}?si",
            f"https://open.spotify.com/track/{track_id}?",
            f"https://open.spotify.com/track/{track_id}",
            f"http://open.spotify.com/track/{track_id}/",
            f"spotify:track:{track_id}",
            track_id,
        ]
        for identifier in valid_identifiers:
            with self.subTest(identifier=identifier):
                url = validate_spotify_url(identifier)
                self.assertEqual(url, f"https://open.spotify.com/track/{track_id}")

    def test_invalid(self) -> None:
        track_id = "1UUBkjBgce8P53hQC3nHtU"
        invalid_identifiers = [
            f"htt://open.spotify.com/track/{track_id}?si=76e0e46d5d7742e0",
            f"https:/open.spotify.com/track/{track_id}?si=",
            f"https://spotify.com/track/{track_id}?si",
            f"https://open.spotify.com/{track_id}?",
            "https://open.spotify.com/track/",
            "https://open.spotify.com/track?si=76e0e46d5d7742e0",
            f"track:{track_id}",
            track_id + "/",
        ]
        for identifier in invalid_identifiers:
            with self.subTest(identifier=identifier):
                with self.assertRaises(ValidationError):
                    validate_spotify_url(identifier)


class TestBuildTrackUrl(unittest.TestCase):
    """Test :py:func:'main.build_track_url'"""

    def test_build(self) -> None:
        url = build_track_url("1UUBkjBgce8P53hQC3nHtU")
        self.assertEqual(url, "https://open.spotify.com/track/1UUBkjBgce8P53hQC3nHtU")


class TestGetUrl(unittest.TestCase):
    """Test :py:func:'main.get_url'"""

    def test_url_exists(self) -> None:
        event = make_api_gateway_event("https://open.spotify.com/track/1UUBkjBgce8P53hQC3nHtU")
        url = get_url(event)
        self.assertEqual(url, "https://open.spotify.com/track/1UUBkjBgce8P53hQC3nHtU")

    def test_url_missing(self) -> None:
        event = deepcopy(API_GATEWAY_EVENT)
        del event["queryStringParameters"]["url"]
        url = get_url(event)
        self.assertEqual(url, "")

    def test_query_params_missing(self) -> None:
        event = deepcopy(API_GATEWAY_EVENT)
        del event["queryStringParameters"]
        url = get_url(event)
        self.assertEqual(url, "")


class TestGetMeta(unittest.TestCase):
    """Test :py:func:`main.get_metadata`"""

    def test_valid_html(self) -> None:
        meta_props = get_meta_props(SPOTIFY_TRACK_HTML_CONTENT)
        self.assertEqual(meta_props, EXPECTED_META_PROPS)

    def test_valid_html_without_meta(self) -> None:
        meta_props = get_meta_props(SPOTIFY_TRACK_HTML_CONTENT_WITHOUT_META)
        self.assertEqual(meta_props, {})

    def test_invalid_html(self) -> None:
        meta_props = get_meta_props("This is not a valid HTML")
        self.assertEqual(meta_props, {})


class TestGetSongData(unittest.TestCase):
    """Test :py:func:`main.get_song_data`"""

    def test_full_props(self) -> None:
        song_data = get_song_data(EXPECTED_META_PROPS)
        expected_song_data = {
            "title": "Nothing is Safe",
            "description": "clipping. · Song · 2019",
            "audio_url": "https://p.scdn.co/mp3-preview/78868f138472a9b6ba194fefafb536a7b017fc57",
            "image_url": "https://i.scdn.co/image/ab67616d0000b2734d1c35edb0aafeb16ef7c7df",
            "url": "https://open.spotify.com/track/52UthFOfrbtnjBMK6jyLqk",
        }
        self.assertEqual(song_data, expected_song_data)

    def test_empty_props(self) -> None:
        song_data = get_song_data({})
        expected_song_data = {
            "title": "",
            "description": "",
            "audio_url": "",
            "image_url": "",
            "url": "",
        }
        self.assertEqual(song_data, expected_song_data)


class TestGetResponse(unittest.TestCase):
    """Test :py:func:`main.get_response`"""

    def test_get_response_successfully(self) -> None:
        request = make_api_gateway_event("https://open.spotify.com/track/1UUBkjBgce8P53hQC3nHtU")
        mock_response = MagicMock(content=b'<meta property="og:title" content="Nothing is Safe"/>')
        with mock.patch("requests.get", return_value=mock_response):
            response = get_response(request)

        self.assertIsInstance(response, dict)
        self.assertIn("statusCode", response)
        self.assertIn("body", response)
        self.assertEqual(response["statusCode"], 200)

    def test_get_response_no_data_in_content(self) -> None:
        request = make_api_gateway_event("https://open.spotify.com/track/1UUBkjBgce8P53hQC3nHtU")
        with mock.patch("requests.get", return_value=MagicMock(content=b"")):
            with self.assertRaises(NoDataError):
                get_response(request)

    @mock.patch("main.get_url")
    @mock.patch("main.validate_spotify_url")
    @mock.patch("main.get_content")
    @mock.patch("main.get_meta_props")
    @mock.patch("main.get_song_data")
    @mock.patch("main.json.dumps")
    def test_get_response_all_calls_in_order(
        self,
        mock_json_dumps,
        mock_get_song_data,
        mock_get_meta_props,
        mock_get_content,
        mock_validate_spotify_url,
        mock_get_url,
    ) -> None:
        request = make_api_gateway_event("ABC")
        get_response(request)

        mock_get_url.assert_called_with(request)
        mock_validate_spotify_url.assert_called_with(mock_get_url.return_value)
        mock_get_content.assert_called_with(mock_validate_spotify_url.return_value)
        mock_get_meta_props.assert_called_with(mock_get_content.return_value.decode.return_value)
        mock_get_song_data.assert_called_with(mock_get_meta_props.return_value)
        mock_json_dumps.assert_called_with(mock_get_song_data.return_value)


class TestHandleError(unittest.TestCase):
    def test(self) -> None:
        error = ValidationError("Error")
        ret_val = handle_error(error)
        self.assertEqual(
            ret_val,
            {
                "statusCode": 400,
                "headers": {
                    "Access-Control-Allow-Headers": "Content-Type",
                    "Access-Control-Allow-Origin": "*",
                    "Access-Control-Allow-Methods": "OPTIONS,GET",
                },
                "body": json.dumps({"detail": "Error"}),
            },
        )


class TestLambdaHandler(unittest.TestCase):
    @mock.patch("requests.get")
    def test_success(self, mocked_get) -> None:
        # Mocking requests
        response_mock = Response()
        response_mock.status_code = 200
        response_mock._content = SPOTIFY_TRACK_HTML_CONTENT.encode("utf-8")
        mocked_get.return_value = response_mock

        event = make_api_gateway_event("https://open.spotify.com/track/1UUBkjBgce8P53hQC3nHtU")
        response = lambda_handler(event, {})
        self.assertIsInstance(response, dict)
        self.assertIn("statusCode", response)
        self.assertIn("body", response)
        self.assertEqual(response["statusCode"], 200)

    def test_error(self) -> None:
        errors = [ValidationError("Error message"), NoDataError("Error message")]
        for error in errors:
            with self.subTest(error=error.__class__.__name__):
                with mock.patch(
                    "main.get_response",
                    side_effect=lambda *args, **kwargs: raise_error(error),
                ):
                    response = lambda_handler(mock.MagicMock(), {})
                self.assertIsInstance(response, dict)
                self.assertIn("statusCode", response)
                self.assertIn("body", response)
                self.assertEqual(response["statusCode"], 400)
                self.assertEqual(response["body"], json.dumps({"detail": "Error message"}))

    def test_no_url_error(self) -> None:
        event = make_api_gateway_event(None)  # type: ignore
        response = lambda_handler(event, {})
        self.assertIsInstance(response, dict)
        self.assertIn("statusCode", response)
        self.assertIn("body", response)
        self.assertEqual(400, response["statusCode"])


if __name__ == "__main__":
    unittest.main()


test.sh
#!/usr/bin/env bash

if [[ -z "${VIRTUAL_ENV}" ]]; then
  echo "Virtualenv must be activated"
  exit 1
else
  if [ ! -f .env ]
  then
    export $(cat .env | xargs)
    python -m unittest src/test.py
  else
    python -m unittest src/test.py
  fi
fi

src/utils.py
import re

ul = "\u00a1-\uffff"  # Unicode letters range (must not be a raw string).

# IP patterns
ipv4_re = r"(?:25[0-5]|2[0-4]\d|[0-1]?\d?\d)(?:\.(?:25[0-5]|2[0-4]\d|[0-1]?\d?\d)){3}"
ipv6_re = r"\[[0-9a-f:.]+\]"  # (simple regex, validated later)

# Host patterns
hostname_re = r"[a-z" + ul + r"0-9](?:[a-z" + ul + r"0-9-]{0,61}[a-z" + ul + r"0-9])?"
# Max length for domain name labels is 63 characters per RFC 1034 sec. 3.1
domain_re = r"(?:\.(?!-)[a-z" + ul + r"0-9-]{1,63}(?<!-))*"
tld_re = (
    r"\."  # dot
    r"(?!-)"  # can't start with a dash
    r"(?:[a-z" + ul + "-]{2,63}"  # domain label
    r"|xn--[a-z0-9]{1,59})"  # or punycode label
    r"(?<!-)"  # can't end with a dash
    r"\.?"  # may have a trailing dot
)
host_re = "(" + hostname_re + domain_re + tld_re + "|localhost)"


URL_REGEX = re.compile(
    r"^(?:[a-z0-9.+-]*)://"  # scheme is validated separately
    r"(?:[^\s:@/]+(?::[^\s:@/]*)?@)?"  # user:pass authentication
    r"(?:" + ipv4_re + "|" + ipv6_re + "|" + host_re + ")"
    r"(?::\d{2,5})?"  # port
    r"(?:[/?#][^\s]*)?"  # resource path
    r"\Z",
    re.IGNORECASE,
)


URI_REGEX = re.compile(r"spotify://track(?:[/?#][^\s]*)?\Z", re.IGNORECASE)


def remove_query_params(url: str) -> str:
    return url.split("?")[0]


def is_url(value: str) -> bool:
    return bool(URL_REGEX.match(value) or URI_REGEX.match(value))


main.py
import json
import os
import re
from collections import defaultdict
from typing import Any, Dict, List, TypedDict, Union

import requests
from bs4 import BeautifulSoup

from utils import is_url, remove_query_params


class ValidationError(RuntimeError):
    pass


class NoDataError(RuntimeError):
    pass


SongData = TypedDict(
    "SongData",
    {"title": str, "description": str, "audio_url": str, "image_url": str, "url": str},
)
MetaProps = Dict[str, List[str]]
RequestType = Dict[str, Any]
ResponseType = TypedDict("ResponseType", {"statusCode": int, "headers": Dict[str, Any], "body": str})


def response(status_code: int, body: Any) -> ResponseType:
    return {
        "statusCode": int(status_code),
        "headers": {
            "Access-Control-Allow-Headers": os.environ.get("ACCESS_CONTROL_ALLOW_HEADERS", ""),
            "Access-Control-Allow-Origin": os.environ.get("ACCESS_CONTROL_ALLOW_ORIGIN", ""),
            "Access-Control-Allow-Methods": os.environ.get("ACCESS_CONTROL_ALLOW_METHODS", ""),
        },
        "body": body if isinstance(body, str) else json.dumps(body),
    }


def get_url(request: RequestType) -> str:
    try:
        return request["queryStringParameters"]["url"]
    except KeyError:
        return ""


def build_track_url(track_id: str) -> str:
    url_template = "https://open.spotify.com/track/{track_id}"
    return url_template.format(track_id=track_id)


def validate_spotify_url(url: str) -> str:
    """
    URL might be one of those:
    1. Spotify track URL (https://open.spotify.com/track/<track-id>)
    2. Spotify track URI (spotify:track:<track-id>)
    3. Track ID itself
    """
    if not url:
        raise ValidationError("'url' parameter required")

    url_regex = re.compile(r"^https?://" r"open\.spotify\.com/track/" r"(?P<track_id>[A-Za-z0-9]+)" r"/?\??\S*$")
    uri_regex = re.compile("^spotify:track:(?P<track_id>[A-Za-z0-9]+)$")
    track_id_regex = re.compile("^(?P<track_id>[A-Za-z0-9]+)$")

    for regex in [url_regex, uri_regex, track_id_regex]:
        match = regex.match(url)
        if match:
            return build_track_url(match.groupdict()["track_id"])

    raise ValidationError("Invalid track identifier. It must be URL, URI or ID")


def get_content(url: str) -> bytes:
    resp = requests.get(url)
    resp.raise_for_status()
    return resp.content


def get_meta_props(content: str) -> MetaProps:
    html = BeautifulSoup(content, features="html.parser")
    props = defaultdict(list)
    for tag in html.find_all("meta"):
        try:
            prop = tag["property"]
            content = tag["content"]
        except KeyError:
            pass
        else:
            if is_url(content):
                content = remove_query_params(content)
            props[prop].append(content)
    return props


def get_song_data(meta_props: MetaProps) -> SongData:
    def prep_value(value: Union[List[str], str]) -> str:
        if isinstance(value, list):
            if len(value) == 1:
                return value[0]
        return str(value)

    return {
        "title": prep_value(meta_props.get("og:title", "")),
        "description": prep_value(meta_props.get("og:description", "")),
        "audio_url": prep_value(meta_props.get("og:audio", "")),
        "image_url": prep_value(meta_props.get("og:image", "")),
        "url": prep_value(meta_props.get("og:url", "")),
    }


def get_response(request: RequestType) -> ResponseType:
    url = get_url(request)
    url = validate_spotify_url(url)
    content = get_content(url)
    meta_props = get_meta_props(content.decode("utf-8"))
    if not meta_props:
        raise NoDataError("Data not found in Spotify response")

    song_data = get_song_data(meta_props)

    return response(200, song_data)


def handle_error(error: Exception) -> ResponseType:
    error_body = {"detail": str(error)}
    return response(400, error_body)


def lambda_handler(event: RequestType, context: Any) -> ResponseType:
    try:
        return get_response(event)
    except (ValidationError, NoDataError) as e:
        return handle_error(e)
